### PR TITLE
feat: Fetch Payment Terms from linked Sales/Purchase Order

### PR DESCRIPTION
**Pull request by GangaManoj**
Backport of https://github.com/frappe/erpnext/pull/26397

`no-docs`

### CODE REVIEW

File: erpnext/stock/doctype/purchase_receipt/test_purchase_receipt.py [hunk 1]
```diff
@@ -1052,6 +1052,58 @@ def test_service_item_purchase_with_perpetual_inventory(self):
 
 		frappe.db.set_value('Company', company, 'enable_perpetual_inventory_for_non_stock_items', before_test_value)
 
+	def test_purchase_receipt_with_exchange_rate_difference(self):
```
Comment (deepeshgarg007) [inline erpnext/stock/doctype/purchase_receipt/test_purchase_receipt.py]: Remove this test

File: erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py [hunk 2]
```diff
@@ -1142,6 +1142,36 @@ def set_status(self, update=False, status=None, update_modified=True):
 		if update:
 			self.db_set('status', self.status, update_modified = update_modified)
 
+# to get details of purchase invoice/receipt from which this doc was created for exchange rate difference handling
+def get_purchase_document_details(doc):
```
Comment (deepeshgarg007) [inline erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py]: Remove this

File: erpnext/buying/doctype/purchase_order/test_purchase_order.py [hunk 3]
```diff
@@ -632,7 +632,7 @@ def test_po_for_blocked_supplier_payments_past_date(self):
 			else:
 				raise Exception
 
-	def test_terms_does_not_copy(self):
+	def test_terms_are_not_copied_if_automatically_fetch_payment_terms_is_unchecked(self):
```
Comment (deepeshgarg007) [inline erpnext/buying/doctype/purchase_order/test_purchase_order.py:639]: The order of test case execution is not fixed. There may be cases when another test case that is enabling the setting may get run first hence this test is failing. Disable the setting agiain in other test case once the test is complete

File: erpnext/selling/doctype/sales_order/test_sales_order.py [hunk 4]
```diff
@@ -1229,7 +1229,42 @@ def test_so_cancellation_when_si_drafted(self):
...
+
+		self.assertEqual(so.payment_terms_template, si.payment_terms_template)
+		compare_payment_schedules(self, so, si)
+
+def automatically_fetch_payment_terms(enable=1):
+	accounts_settings = frappe.get_doc("Accounts Settings")
+	accounts_settings.automatically_fetch_payment_terms = enable
+	accounts_settings.save()
+
+def compare_payment_schedules(doc, doc1, doc2):
```
Comment (deepeshgarg007) [inline erpnext/selling/doctype/sales_order/test_sales_order.py:1258]: Use ORM for queries wherever possible

**Review by deepeshgarg007 — CHANGES_REQUESTED** [review]

File: erpnext/stock/doctype/purchase_receipt/test_purchase_receipt.py (hunk 1 above)
Comment (GangaManoj) [reply erpnext/stock/doctype/purchase_receipt/test_purchase_receipt.py]: Done.

File: erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py (hunk 2 above)
Comment (GangaManoj) [reply erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py]: Done.

File: erpnext/buying/doctype/purchase_order/test_purchase_order.py (hunk 3 above)
Comment (GangaManoj) [reply erpnext/buying/doctype/purchase_order/test_purchase_order.py:639]: Done.

File: erpnext/selling/doctype/sales_order/test_sales_order.py (hunk 4 above)
Comment (GangaManoj) [reply erpnext/selling/doctype/sales_order/test_sales_order.py:1258]: Done.

**Review by deepeshgarg007 — APPROVED** [review]

**Pull request merged**